### ISSUE TITLE
Implement Game Model

### ISSUE THREAD

**Issue by zjcers** [EECS448-KitchenUtensils/DefinitelyNotProject3#7]

**Comment by zjcers** [comment]
This is happening concurrently with #5 and #6 as I figure out chess.

**Comment by zjcers** [comment]
Mostly done. The PossibleMoves method on each piece is getting updated to take in an interface that allows lookups by BoardPosition so that PossibleMoves can be perfectly accurate. Previously, this was going to be delegated to the ChessBoard class, but proved to be a hilariously bad idea. #6 will be updated ex post facto.

**Comment by jackdweber** [comment]
Interface

* PossibleMoves
  * Arguments: Piece coordinates
  * Returns: Enumerable of (BoardPosition, MoveType) that would be valid moves
  * Side-effects: None
* SpeculateCheckMate
  * Arguments: Piece current position, piece speculative position
  * Returns: true if this move would cause a checkmate against an enemy king
  * Side-effects: None
* GetPieces
  * Arguments: None
  * Returns: Enumerable of all pieces on board
  * Side-effects: None
* GetPieceByPosition
  * Arguments: Position
  * Returns: A ChessPiece or null
  * Side-effects: None
* MakeMove
  * Arguments: Piece current position, piece desired position
  * Returns: MoveType (Failure, Capture, Move)
  * Side-effects: Changes turn to point at next player, possibly mutates the board state
* GetTurn
  * Arguments: None
  * Returns: Current player
  * Side-effects: None
* Events:
  * PieceMoved - Fired when a move is accepted but does not result in piece capture
  * PieceCaptured - Fired when a move causes a piece to be captured
  * NextTurn - Fired when the active player changes

**Comment by zjcers** [comment]
Don't forget that pawns can't move backwards, but what backwards is depends on the player.

**Comment by zjcers** [comment]
Nearing completion. Only need to implement end of game detection and making moves (though the latter should be trivial).

**Comment by zjcers** [comment]
Only end of game detection is missing. Merged branch server-master into master so that client is unblocked.

**Comment by zjcers** [comment]
Pushing end of game detection into another issue since that's separate functionality
